### ISSUE TITLE
This Qualifier in Knockout Foreach

### ISSUE THREAD

**Issue by DVaughan** [nikhilk/scriptsharp#203]
I'm loving the way I can bind my Script# viewmodels using Knockout. 
I'm using a Knockout data-binding to display a collection of model objects that are located in an ObservableArray in a viewmodel. Rather like this:
&lt;table>
   &lt;tbody data-bind="foreach: people"&gt;
        &lt;tr&gt;&lt;td&gt;&lt;button data-bind="click: $parent.update"&gt;Update&lt;/button&gt;&lt;/td&gt;&lt;/tr&gt;
...
Notice that I'm using the $parent qualifier to call a method on the outer viewmodel and not on the 'Person' object. This works so far, however, in my viewmodel the Update method calls a base class (ViewModelBase) method called SetTitle. SetTitle updates an Observable<string> title value. It does so using the 'this' qualifier which resolves to the current 'Person' object. It's rather odd. I'm not sure what to do about this one. I've replicated the behavior and it doesn't appear to be particular to this case. Any work around would be greatly appreciated.

**Comment by nikhilk** [comment]
Is this a script# issue, or a knockout issue?

What is the expected script? And which part is not possible/incorrectly generated?

**Comment by DVaughan** [comment]
Hi Nikhil,

My feeling is that what is occurring is a case of binding loss and that the Script# output should use a different mechanism for resolving the specified member method, such as by outputting 'apply' or 'call' statements in place of 'this'. 'this' does not produce the expected behavior under some circumstances. From what I've seen this includes nested scopes in Knockout foreach loops. 
You may already be familiar with it but I found an article about this here. http://www.alistapart.com/articles/getoutbindingsituations 

Cheers,
Daniel

**Comment by nikhilk** [comment]
If "this" isn't resolving to the right object in a function that was passed to something that calls the function in context of some other object as the "this", then I'd say the problem isn't really in how the script is generated to invoke a member method or a base member method. Rather, I'd say the way to fix the problem is really to pass a delegate where the function was handed off ... the delegate captures the binding between a function, and an object context, and does what you're saying - using call/apply to restore the this context to the object you expected.

I don't know the internals of how knockout works to talk to it specifically ... hence sharing actual code, expected output all helps, but you can see other examples of the same scenario. For example, when you call something as simple as addEventListener, the function you pass in will by default be called with the global object as the this, but creating a delegate sets up the binding between the function and an object instance. You'll need to figure out the same in this particular scenario - how to pass a delegate that has captured the function and object context.

Hope that helps.

Update: as I once more look at the little snippet you've got ... I see $parent.update. Perhaps what knockout is doing is simply looking up the update member on an object and invoking it directly. I would have expected knockout to invoke the update method passing the $parent object as the this context ... no idea if it does or doesn't. If it doesn't, and that is the right behavior for it, then you'll need to compensate for the lack of that by creating a member that holds your delegate, and invoking that Rough code...

this.xyz = Update; // will create a delegate

data-bind="click: $parent.xyz"

**Comment by DVaughan** [comment]
Hi Nikhil,

I've put together a simple project demonstrating the issue.
https://docs.google.com/open?id=0B-ThP21bMCrhekJRekJGTHpTNEk
(select File/Download)

In the sample, the index.html page is bound to the MainViewModel. The MainViewModel has a method called PerformSomeAction and an ObservableArray of Person objects. A foreach data-binding expression displays each Person, while within the foreach a button is bound to the ViewModel's PerformSomeAction method. When clicking on the button the PerformSomeAction method is correctly called, but accessing any of its other members does not work because 'this' resolves the current Person and not the MainViewModel.

I'm using Script# as a means to write the business logic for my application. Knockout is the glue that allows me to marry the Script# output with my pages and custom JS. Apart from this issue, it works well.

Cheers,
Daniel

**Comment by nikhilk** [comment]
So I really can't tell ... your description makes it sound like a Knockout issue... the fact that it is using the person object as the this context, rather than parent, even though you have that in your markup.

If its not a Knockout issue, you'll need to provide more information to make this issue actionable... what is the code you'd expect to write if you were to code javascript directly, such that Knockout works and does what you expect it to do? And then look at the delta to figure out why the delta and what to do about it. Makes sense?

**Comment by DVaughan** [comment]
Okay, so here is a candidate solution.
https://docs.google.com/open?id=0B-ThP21bMCrhSTNiT2hGbHEtU3M
(File/Download)

In my opinion it's not particular to Knockout, it's just how 'this' works in Javascript as I mentioned in an earlier post. 'this' is resolved relative to its context. I've seen others using a local variable assigned to 'this' when the object is initialized. That is the approach I took in the candidate solution. I've used a couple of script literals to demonstrate that it works. Unless someone knows a better approach, I suggest that 'self' be used in place of this in the Script# compiler output.

**Comment by nikhilk** [comment]
Can't really go over a whole project right now ... at least not until the weekend. Its much easier to look at code snippets here, email etc. and respond in between things.

That said, as I mentioned to the previous reply, that is the whole point of a delegate in c# ... it binds a method to an instance, so you don't have to manually implement the this capturing via a self local. However if Knockout requires you to capture the this context yourself, then you should certainly be able to do so without a Script.Literal.

```
MyClass self = this;
```

should work? right?

**Comment by DVaughan** [comment]
Okay, please take a look at the weekend. The project contains two projects; a script# class library and a web project. The script# proj has one file worth looking at MainViewModel.cs. The web project has one file worth looking at: index.html.

"MyClass self = this" does not work because the Script# compiler outputs 'this._self'. Script literals are required.

**Comment by michaelaird** [comment]
Daniel,

We ran into this issue with knockout as well. If you change your binding handler to this:

data-bind="click: function() { $parent.performSomeAction(); }"

that will keep "this" properly scoped to what you expect.

Or, if you need a reference to the current item that was clicked, use this:

data-bind="click: function() { $parent.performSomeAction(this); }"

and have PerformSomeAction take a parameter of the object type that was bound. 

If this isn't clear enough, let me know and i can update your sample project to demonstrate.

Michael.

**Comment by michaelaird** [comment]
Another alternative that we've used is to declare your click handler as a public Action<T> on your viewmodel. In the constructor, assign it a delegate which will properly capture "this". like so 

``` c#
        public Action<Person> PerformSomeAction2;

        public MainViewModel()
        {
            PerformSomeAction2 = delegate(Person p)
            {
                ShowDialog();
            };

            //Script.Literal("_self = this");
            self = this;

            for (int i = 0; i < 5; i++)
            {
                Person person = new Person("Example Person " + i);
                People.Push(person);
            }
        }
```

Again, if this doesn't make sense, i can send you a code sample.

**Comment by michaelaird** [comment]
Daniel,

I don't think your sample wrapped the call in an anonymous function. That's
the key.

I'm on the streetcar right now so I can't provide a full sample but I'll
put one together when I get home.
On Jul 17, 2012 7:28 PM, "DVaughan" <
reply@reply.github.com>
wrote:

**Comment by DVaughan** [comment]
Michael, wrapping the binding in a function is a reasonable work around. I would still love to see a fix integrated into script#. I imagine this would pop up under other circumstances since 'this' doesn't behave the same way in Javascript as it does in C#.

**Comment by michaelaird** [comment]
I do agree that it would be nice to have a more "permanent" reference to
'this' but I think that's something we have to live with in JS.

Another architectural approach is to be more Functional and make your click
handlers static and have them take and operate on parameters.
On Jul 17, 2012 7:33 PM, "DVaughan" <
reply@reply.github.com>
wrote:

**Comment by DVaughan** [comment]
I don't think it's something that has to be lived with in C# though, since you can capture the object using a field and it's a change to the compiler output. Changing the style of development to accommodate what is essentially a bug isn't the right approach, and static handlers require needless plumbing code... matching up arguments and so forth.

**Comment by nikhilk** [comment]
Good discussion here. Lets step back for a moment and look at how things work.

There are two scenarios in general in the context of invoking functions as it relates to this discussion:
1. You are writing code that hands a reference to some function to another piece of code, which will then invoke the function at some point, and when it does, the "this" context should be set to the current object handing out the function reference. Example - addEventListener. When you write this code:
   
   someElement.AddEventListener("event", OnEvent);

the compiler will generate code to create a delegate and pass that to addEventListener. Something like:

```
someElement.addEventListener('event', ss.Delegate.create(this, this.onEvent));
```

the delegate is a function that when invoked, turns around to call the onEvent function with the right this context.
1. Some code is looking up a function member of an instance object, and then invoking that function, i.e. the scenario we are discussing here. In this case the caller is the one that knows about both object instance and function reference. Only it can bind the two together when it actually invokes the function, by doing something like:
   
   functionInstance.call(objectInstance, args)

I think Knockout doesn't seem to be doing the right thing. When the markup has $parent.performSomeAction, I'd expect that Knockout would lookup the $parent object instance for a member named performSomeAction, and then call it, passing in $parent as the this object, because that is normal script behavior. This is why the workaround suggested above works - its compensating for what knockout should be, but isn't doing.

```
data-bind="click: function() { $parent.performSomeAction(this); }"
```

The alternative if you don't want to workaround the unexpected knockout behavior is to expose a delegate in your view model, i.e. turn this scenario 2 problem into a scenario 1 case is by using the other workaround suggested.

```
public Action<Person> PerformSomeAction2;

PerformSomeAction2 = delegate(Person p) { ShowDialog(); };
```

Either way you'll need to workaround what Knockout doesn't seem to be doing.

The script# compiler is doing the right binding of object instance and function reference when you ask it do so, i.e. when handing out a function to some other code.

So I don't see some fix coming in script# for this.

**Comment by michaelaird** [comment]
Here's a comment from RPN confirming that this is an oddity/limitation of Knockout:

https://groups.google.com/forum/?fromgroups#!searchin/knockoutjs/click$20this/knockoutjs/mxpkJ_YOh7o/LAL0liXQcEsJ

**Comment by DVaughan** [comment]
Thanks for your response Nikhil. I hope that at some point this will be resolved. I would like to see the fundamentals, like the resolution of 'this', preserved by Script#, no matter what other third party libraries do.

Cheers,
Daniel

**Comment by nikhilk** [comment]
@DVaughan ... hoping my response made it clear that script# does in fact do that.

**Comment by DVaughan** [comment]
Since 'this' does not behave like C#, using 'this' in the compiler output seems like the wrong choice to me. A 'self' variable would make the output more robust and produce reliable behavior under all conditions. But ho hum, let's move on. I see you've closed the issue anyway.

**Comment by nikhilk** [comment]
There is some thought on use of self to be able to capture the this object reference for use within a closure (specifically anonymous method) rather than use a full delegate for them (in the future script generation logic). However that won't help this case.

I don't agree that "this" can't behave as you expect. Really in this case, the Knockout design is to invoke a function against the context of the item in the collection it is enumerating, rather than calling it against the object from which the method was looked up to match the semantics of the markup (which IMO is just plain odd ... as reflected in the google groups link above) and as such, I don't see there being anything the compiler can do, without starting to consider building in framework-specific constructs and quirks into the compiler.

**Comment by DVaughan** [comment]
The sample project linked to above solves the issue by using a 'self' variable via Script.Literals.

Yes, the Knockout behavior is wonky, but in my view Script# should be producing JS that ensures the fidelity of the source C#.
